### ISSUE TITLE
The library is not thread safe. 

### ISSUE THREAD

**Issue by skyguy94** [jmp75/rdotnet#7]
Calling it in concurrent threads is likely to crash, even if user synchronizes the calls at higher level. The lock in Preserve/Unpreserve functions doesn't really help because: 1. it can be called in gc threads concurrently with other R API functions since they are not locked, both R API and the GetFunction() utility are not thread safe; 2. if user explictly calls Unpreserve, R_ReleaseObject can be invoked second time since the lock is inside the flag checking.

I think the solution is to use engine-wide lock to synchronize every R API invocation. The locking overhead should be neglectable comparing to the expensive R API invocations. I've tested with this change and I think the library is thread safe now. It can pass the MultiThreadingTest and StressTest with no problem. The hack on CStackLimit can also be dropped after this change. I can push the change to some branch if someone is interested.

**Comment by skyguy94** [comment]
I'd love to see your work. I was never really thrilled with the CStackLimit hack, but it doesn't seem to cause real problems.

**Comment by hyh** [comment]
I've made the change available at https://github.com/hyh/rdotnet/tree/thread-safe
Note:
1. the code is unpolished and for demonstrating the idea only
2. I didn't make all the functions thread-safe, just the ones used by MultiThreadTest to make it work.
3. I messed up with some other changes, such as commenting out several ProtectedPointer. please ignor them.

Let me know if you have any comments. Thanks.

**Comment by skyguy94** [comment]
I agree with the concept for thread safety, but one of my issues with the way the R API is imported is that I believe there should be a C# wrapper where all the R code is located. Right now, R.Net is calling R functions willy nilly. It makes sense to me that the flat R api should have a flat pinvoke layer equivalent. The rest of the classes (nee..the synthetic object model) should call that C# layer.

This is precisely what I did in the client/server code: https://github.com/jmp75/rdotnet/blob/client_server/RDotNet.Server/Services/RLanguageService.cs

I think if you have that layer, you can push all the lock code into that one place. I'm working on PR for it, so we can move to a client/server approach that benefits everyone's needs, see (http://rdotnet.codeplex.com/discussions/572139)[Codeplex discussion for multiple concurrent R sessions]

Since the R engine blocks during a call, I think blocking the "unit of work" for a R call is a good approach to ensuring thread-safety in C#

**Comment by hyh** [comment]
I not a fan of the GetFunction() staffs either. But seems like you are creating a C-S service layer on top of that, instead of refactoring it? Is this layer also necessary for in-proc lib users?

I read the codeplex discussion, sorry I can't comment on your C-S architecture. This is not my scenario.
I do agree that SafeHandle is overkill. Simple disposable pattern should be enough for the SEXPs.

**Comment by skyguy94** [comment]
Its really just a shim, a minor extra level of indirection that consolidates R code and platform specific identities in one place. The current implementation is written with WCF and I tossed the platform code. I'll keep using that one internally for my employer, but I want to write a new one that's just sockets and json, similar to the way RServe and RStudio do things. That should keep the cross-platform code working. 

As for in-proc vs out-of-proc, I can currently switch between the two modes (actually three if you include distributed), but I think a good first target is to have a single process running on the backend, again, like RStudio.

**Comment by pfgherardini** [comment]
Hi all,

Please excuse me if this issue has already been addressed elsewhere, I looked around but I wasn't able to find a definitive answer. Is there a way to have a multi-threaded application where each thread fires up its own REngine, does its calculations and then disposes of it? My use case is a webserver where users need to be able to launch independent (and expensive) R calculations. I don't care at all about the cost of having multiple REngines (if that's even possible) because any startup/initialization time is going to be insignificant compared to the cost of the actual computation.

Thanks in advance for your help,

Federico

**Comment by jmp75** [comment]
@pfgherardini There can only be one R engine in a process, and R itself is largely not thread-safe.

**Comment by lsusatyo-gcmlp** [comment]
Hi Ritch,

I noticed that you have a fork on this rdotnet project. Is this done to introduce the lock mechanism on the engine level? If so, any plan to consolidate this feature? Currently we are using the rdotnet and are trying to find ways to resolve the instability issue of multithreading.

**Comment by dariogriffo** [comment]
Any progress on this?
